Modify-able config.

otherdaniel opened the pull request:
[This isn't quite ready yet. Using a PR for tool support and discussion.]

This re-writes the config logic, based on discussions in the recent meetings, and with the intent to address issues #228, #229, and #233.

The corner stones are:

- The Sanitizer (and thus its config) become mutable. It has methods to add stuff to  any of the allow or deny lists.
- The actual config processing is now defined in terms of those methods.
- There is only one "safe-ify" operation, and it gets called once (for safe methods). This is the only different between safe and unsafe versions.
  - (Well, this is only a half-truth. The handling of javascript:-navigation URLs remains an odd-ball that is magically different between "safe" and "unsafe" versions.)
- The configs have an "other" setting, which allows them to express "everything" and "nothing" filters, which in turn allows us to treat the unsafe defaults as just another regular case.
- The entire "known" logic is gone. But in turn, the "baseline" is now back, and is a deny-list.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/otherdaniel/purification/pull/237.html" title="Last updated on Nov 28, 2024, 5:42 PM UTC (e05db5a)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/sanitizer-api/237/85d25b5...otherdaniel:e05db5a.html" title="Last updated on Nov 28, 2024, 5:42 PM UTC (e05db5a)">Diff</a>

mozfreddyb commented:
We discussed this in a synchronous meetings. I'm copying the notes verbatim so we can continue async.

> 
> - Document.parseHTML / Document.parseHTMLUnsafe - How does it handle safety?
> - We should not need `otherMarkup`
> - get/getUnsafe should not be a method
> - Some method for turning unsafe into safe
> - Don’t want to call it “safe”, because we don’t know what safe is
> - Eg. “WithBaseline” (baseline is applied to config)
> - Element vs allowAttribute?
> - setComment why set-method instead of an attribute-setter?
> - Should we use allow instead of allowElement (or element)
> - Anne: How to undo replaceWithChildren
> - Tom: element() currently removes from the replaceWithChildren list
> - Guillaume: “Other name for something captured by allow or block list”
> - We have a baseline know
> - Annevk: baseline remove all unsafe stuff
> - Default “safe list” of allowed elements
> - Annevk: Expose the default list with a static method
> - Anne: Name the method removeUnsafe to make a config safe (instead of safeify, because we can’t easily guarantee safeness for all use cases)
> - It shouldn't be replaceWithChildren but replaceWithChildrenElements (as per earlier group consensus)
> - Anne: Rename methods to addElement, addRemoveElement, addReplaceWithChildrenElement, etc. (prefix with add to make it clear there is a modification of the instance)
> - No group concensus for the `add` prefix. We need to discuss this further.
> - Tom: Make setComment a property like comment
> - Anne: This would imply introspection for the properties and not for other like elements, removeElements etc.
> - Anne: Remove getUnsafe, instead a method that makes sanitizer safe (removeUnsafe)
> - Guillaume: get() returns whatever the current state is
> - Anne: Could we use toJSON instead of get?
> - Guillaume/Frederik: We need to worry about having functions that are not JSON compatible in the future (e.g. if we add a fallback feature in v2)
> - all: Let’s keep get then.
>

mozfreddyb reviewed (changes requested):
Really great progress so far. We have some nits (and some larger questions) in the review notes, please see above.

otherdaniel commented:
- Added issue links to issues for naming and other markup
- Added "unsafe" handling to parseHTMLUnsafe.
- Replaced `getUnsafe` query method with `removeUnsafe` modifier.
- Naming mostly deferred. But made sure all method names start with a verb.
- Finished handling of element-dependent attributes.
- Various editorial / link-ification improvements.

mozfreddyb commented on index.bs:
```diff
@@ -217,47 +228,157 @@ dictionary SetHTMLOptions {
 };
 </pre>
 
+ISSUE(238): Final naming TBD.
+
+ISSUE(240): "other markup" TBD.
+
+ISSUE: Should these be setter methods -- particularly the setXXX(boolean) --
+    or setters or properties or somesuch?
+
+ISSUE: Should the modifier methods return a reference to [=this=], so that you
+    can 'chain' methods?
+    (e.g. `sanitizer.allowElement("a").allowElement("span")`).
+
```
Please file issues for those that are only in the file

otherdaniel replied:
Done.

mozfreddyb commented on index.bs:
```diff
@@ -217,47 +228,157 @@ dictionary SetHTMLOptions {
 
-1. Store |config| in [=this=]'s [=internal slot=].
+1. Let |valid| be the return value of [=set a config|Set=] |config| on [=this=].
+1. If not |valid|, throw a {{TypeError}1. If not |valid|, throw a {{TypeError}}}
```
```suggestion
1. If not |valid|, throw a {{TypeError}.
```

otherdaniel replied:
Done.

mozfreddyb commented on index.bs:
```diff
@@ -297,6 +418,7 @@ dictionary SanitizerConfig {
 };
 </pre>
 
+ISSUE: Sould members be required, or have declared defaults?
```
(Same as above, pls file ISSUE for those that area only in the document)

otherdaniel replied:
Done. (Also removed issues that had been resolved, like this one, or otherMarkup.

annevk commented on index.bs:
```diff
@@ -308,40 +430,53 @@ To <dfn>set and filter HTML</dfn>, given an {{Element}} or {{DocumentFragment}}
+<div algorithm>
 For the main <dfn>sanitize</dfn> operation, using a {{ParentNode}} |node|, a
-[=SanitizerConfig/canonical=] {{SanitizerConfig}} |config|, run these steps:
+{{Sanitizer}} |sanitizer| and a [=boolean=] |safe|, run these steps:
```
```suggestion
{{Sanitizer}} |sanitizer|, and a [=boolean=] |safe|, run these steps:
```

otherdaniel replied:
Done.

annevk commented on index.bs:
```diff
@@ -308,40 +430,53 @@ To <dfn>set and filter HTML</dfn>, given an {{Element}} or {{DocumentFragment}}
+{{Sanitizer}} |sanitizer| and a [=boolean=] |safe|, run these steps:
+
+1. Let |config| be the value of |sanitizer|'s [=internal slot=].
+1. If |safe|, let |config| be the result of calling [=remove unsafe=] on |config|.
```
```suggestion
1. If |safe| is true, then set |config| to the result of calling [=remove unsafe=] on |config|.
```

annevk commented on index.bs:
```diff
@@ -358,324 +493,120 @@ For the main <dfn>sanitize</dfn> operation, using a {{ParentNode}} |node|, a
+   {{SanitizerConfig/replaceWithChildrenElements}}, or
+   {{SanitizerConfig/attributes}}.
+1. Let |result| be a copy of |config|.
+1. [=list/For each=] |elem| in
```
Sometimes you use elem, sometimes, item, and sometimes element. I would settle on element for all of these. I'd generally prefer less abbreviations in prose, so also attribute and configuration instead of attr and config.

otherdaniel replied:
Mostly done. I consistently use element, attribute, and configuration now. I still use item now and then, but only for contexts where it could be either an element or an attribute. Happy to rename that, but I couldn't really come up with a nice name for element_or_attribute.

annevk commented on index.bs:
```diff
@@ -358,324 +493,120 @@ For the main <dfn>sanitize</dfn> operation, using a {{ParentNode}} |node|, a
+1. Let |result| be a copy of |config|.
+1. [=list/For each=] |elem| in
+   [=built-in safe baseline config=][{{SanitizerConfig/removeElements}}]:
+    1. Call |result|.{{Sanitizer/removeElement()|removeElement}}(|elem|)
```
You cannot call public API from this algorithm. You instead need to define an internal algorithm that removeElement() and this can both call.

annevk commented on index.bs:
```diff
@@ -358,324 +493,120 @@ For the main <dfn>sanitize</dfn> operation, using a {{ParentNode}} |node|, a
+
+1. [=Assert=]: |config| is a [=dictionary=].
+1. [=list/iterate|For each=] |item| of |config|[{{SanitizerConfig/elements}}] do:
+  1. Call |sanitizer|.{{Sanitizer/allowElement()|allowElement}}(|item|).
```
Same problem as above. You cannot call public API here.

otherdaniel replied:
Done.

otherdaniel commented on index.bs:
```diff
@@ -217,47 +228,157 @@ dictionary SetHTMLOptions {
+<div algorithm>
+The <dfn for="Sanitizer" export>allowElement</dfn>(|element|) method steps are:
+
+1. Let |name| be the result of [=canonicalize a sanitizer name=] |element| with [=HTML namespace=] as the default namespace.
```
Hmm. Canonicalize a sanitizer name is used for both elements and attributes. I could add a "canonicalize an elemnt" and a "canonicale an attribute", both of which would then canonicalize their name, but it seems like that merely adds an indirection. (At least as no other canonicalization steps are performed.)

[Leaving open for now.]

otherdaniel replied:
Done now. In all cases, the algorithms update the argument with the canonicalized version and use only that.

otherdaniel commented on index.bs:
```diff
@@ -217,47 +228,157 @@ dictionary SetHTMLOptions {
+The <dfn for="Sanitizer" export>allowElement</dfn>(|element|) method steps are:
+
+1. Let |name| be the result of [=canonicalize a sanitizer name=] |element| with [=HTML namespace=] as the default namespace.
+1. [=list/Remove=] |name| from [=this=]'s [=internal slot=]'s {{SanitizerConfig/elements}}.
```
Added a [=SanitizerConfig/remove=], to explain removal of a name structure.

Removed "internal slot" in favour of configuration; and systematically added dictionary syntax to reference its members.

(I think I cargo-culted the "internal slot" from ECMAScript, which defines an "internal slot" as a hook that the embedder can use to attach data to a script object. I agree that doesn't actually make any sense over here.)

mozfreddyb commented:
copying over from #238 

change
-   replaceWithChildrenElements to replaceElementWithChildren()
-   setComments() to plural.
 

keep:
-  removeElement()
-   allowElement()
-   removeAttribute()
-   allowAttribute()
-   setDataAttributes()

otherdaniel commented:
Updated the PR with feedback from today's meeting:

- Adopted the updated naming.
- Introduced canonicalization methods for element with attributes; elements; and attributes. Rewrote the caller algorithms so that they canonicalize in the first step, and then only use the canonicalized value.

annevk commented on index.bs:
```diff
@@ -217,48 +227,86 @@ dictionary SetHTMLOptions {
 
 The {{Sanitizer}} configuration object encapsulates a filter configuration.
-The same config can be used with both safe or unsafe methods. The intent is
+The same configuration can be used with both safe or unsafe methods, where
```
```suggestion
The same configuration can be used with both safe and unsafe methods, where
```
Perhaps we should add quotation marks around "safe"?

mozfreddyb replied:
Bikeshed. Sorry: "unsafe and non-unsafe is terrible". How about saying safe and linking to Security Considerations that starts with a bit more of an explicit "This is what we consider safe for our scenario."?

annevk replied:
I'm not sure anyone suggested unsafe and non-unsafe?

mozfreddyb replied:
I suggested & non-suggested that. Nevermind that bit.

I saw your criticism and agree with you that just saying safe without quotes or any extra context could be ambiguous. Putting it in quotes is not a lot of more context either

Considering that, I suggest putting it in quotes and "safe" linking the Security Considerations section would make more sense to me.

otherdaniel replied:
Added a definition to "safe and unsafe" where we first use the terms, and linked there. Added scare quotes, too.

annevk commented on index.bs:
```diff
@@ -217,48 +227,86 @@ dictionary SetHTMLOptions {
 The {{Sanitizer}} configuration object encapsulates a filter configuration.
-The same config can be used with both safe or unsafe methods. The intent is
+The same configuration can be used with both safe or unsafe methods, where
+the safe methods perform an implicit {{removeUnsafe}} operation. The intent is
```
```suggestion
the safe methods perform an implicit {{removeUnsafe}} operation on the passed in configuration and have a default configuration when none is passed. The intent is
```

otherdaniel replied:
Done.

annevk commented on index.bs:
```diff
@@ -217,48 +227,86 @@ dictionary SetHTMLOptions {
 interface Sanitizer {
-  constructor(optional SanitizerConfig config = {});
+  constructor(optional SanitizerConfig configuration = {});
+
```
Let's add a TODO here for `[NewObject] static Sanitizer getDefault()`? (I suppose it could also return a `SanitizerConfig`, but an instance seems better.)

otherdaniel replied:
Done. (Technically, a note instead todo, since that's apparently what bikeshed knows about.)

annevk commented on index.bs:
```diff
@@ -217,48 +227,86 @@ dictionary SetHTMLOptions {
+</div>
 
+<div algorithm>
+The <dfn for="Sanitizer" export>replaceElementWithChildren</dfn>(|element|) method steps are to [=replace an element=] with |element| and [=this=]'s [=Sanitizer/configuration=].
```
Let's also call the internal operation "replace an element with children"?

otherdaniel replied:
Done. Added a pronoun, too, "with its children", since this is just plain text and I think that makes it a little clearer what the operation actually does.

annevk commented on index.bs:
```diff
@@ -308,40 +355,53 @@ To <dfn>set and filter HTML</dfn>, given an {{Element}} or {{DocumentFragment}}
-   |config| and |safe|.
+   Then return  |options|["`sanitizer`"].
+1. [=Assert=]: |options|["`sanitizer`"] is a [=dictionary=].
+1. Return new {{Sanitizer}}(|options|["`sanitizer`"]).
```
Public API.

otherdaniel replied:
Done.

annevk reviewed (approved):
At this point I would be okay with landing this to get the public draft in a better state, but we'll need follow-up issues for `setHTML()` defaulting and such (and any comments in this PR that you decide to postpone addressing).

annevk commented on index.bs:
```diff
@@ -308,40 +361,58 @@ To <dfn>set and filter HTML</dfn>, given an {{Element}} or {{DocumentFragment}}
+1. [=Assert=]: |options| is a [=dictionary=].
+1. If |options|["`sanitizer`"] doesn't [=map/exist=], then:
+   1. Let |result| be a new {{Sanitizer}} instance.
+   1. Call [=set a config=] with an empty [=dictionary=] on |result|.
```
"set a configuration"

otherdaniel replied:
Done.

annevk commented on index.bs:
```diff
@@ -353,330 +424,229 @@ For the main <dfn>sanitize</dfn> operation, using a {{ParentNode}} |node|, a
-default namespace |defaultNamespace|, run the following steps:
+To <dfn for="Sanitizer">set a config</dfn> |configuration| on a {{Sanitizer}} |sanitizer|, do this:
+
+1. [=Assert=]: |configuration| is a [=dictionary=].
```
Instead of asserting this here, you should specify the argument type above.

> To **set a config** given a dictionary |configuration| and a Sanitizer |sanitizer|:

otherdaniel replied:
Done. That indeed reads nicer.

annevk commented on explainer.md:
```diff
@@ -363,6 +363,63 @@ element.setHTML("XXX<!-- Hello world! -->XXX", {sanitizer: config_comments});
+
+The following methods are offered on the Sanitizer object:
+
+- `allow(x, options)`
+  - `options` is an optional dictionary argument.
+    Supported keys are: `"attributes":` and `"removeAttributes":.`
```
```suggestion
- `allowElement(x, options)`
```

otherdaniel replied:
Done. (Also fixed a few more cases in the same chapter.)

otherdaniel commented on index.bs:
```diff
@@ -308,40 +355,53 @@ To <dfn>set and filter HTML</dfn>, given an {{Element}} or {{DocumentFragment}}
-1. Assert: |options|["`sanitizer`"] is either a {{Sanitizer}} instance
+1. [=Assert=]: |options| is a [=dictionary=].
+1. If |options|["`sanitizer`"] doesn't [=map/exist=],
+   then return new {{Sanitizer}}().
```
Done.

otherdaniel commented on index.bs:
```diff
@@ -308,40 +355,53 @@ To <dfn>set and filter HTML</dfn>, given an {{Element}} or {{DocumentFragment}}
+1. Let |configuration| be the value of |sanitizer|'s [=Sanitizer/configuration=].
+1. If |safe| is true, then set |configuration| to the result of calling [=remove unsafe=] on |configuration|.
+1. Call [=sanitize core=] on |node|, |configuration|, and |safe| (as value for
+    handling javascript navigation urls).
```
Done.

otherdaniel commented on index.bs:
```diff
@@ -308,40 +355,53 @@ To <dfn>set and filter HTML</dfn>, given an {{Element}} or {{DocumentFragment}}
+<div algorithm>
+The <dfn>sanitize core</dfn> operation,
+using a {{ParentNode}} |node|, a {{SanitizerConfig}} |configuration|, and a
+[=boolean=] |handle javascript navigation urls|, iterates over the DOM tree
```
Done.

otherdaniel commented on index.bs:
```diff
@@ -353,330 +413,229 @@ For the main <dfn>sanitize</dfn> operation, using a {{ParentNode}} |node|, a
+</div>
+
+<div algorithm>
+To <dfn for="Sanitizer">allow comments</dfn> with |allow| on a {{SanitizerConfig}} |configuration|, do:
```
Done.

otherdaniel commented on index.bs:
```diff
@@ -725,71 +687,75 @@ Equality for [=ordered sets=] is equality of its members, but without
+ISSUE(233): Determine if this actually holds.
+
+
+The <dfn>built-in unsafe default config</dfn> is meant to allow anything.
```
I suspect not. I left this one, since the next step is to re-work the default handling along the lines we discussed in yesterday's meeting, and I hope I can fix all the builtin/default things in one go thhere.

otherdaniel commented on index.bs:
```diff
@@ -725,71 +687,75 @@ Equality for [=ordered sets=] is equality of its members, but without
+    { name: "script", namespace: "http://www.w3.org/2000/svg" }
+  ],
+  removeAttributes: [....],
+  comments: true,
```
Removed.

otherdaniel commented on index.bs:
```diff
@@ -308,40 +361,58 @@ To <dfn>set and filter HTML</dfn>, given an {{Element}} or {{DocumentFragment}}
+1. If |options|["`sanitizer`"] doesn't [=map/exist=], then:
+   1. Let |result| be a new {{Sanitizer}} instance.
+   1. Call [=set a config=] with an empty [=dictionary=] on |result|.
+   1. [=Assert=]: The return value is true.
```
Done.

otherdaniel commented:
I think I resolved all the issues. Will land tomorrow, if no more comments come in tonight.

Next step will be to rework the default handling, and to put up some doc and/or poposal about the built-in lists.

mozfreddyb commented:
🎈
